### PR TITLE
[KAT-956] Changed the difference_type to ptrdiff_t

### PR DESCRIPTION
**Pull request by GioLomia**
I changed the difference type to int64_t for the purposes of resolving comparison issues. I also added some comments to prevent some potential issues that could arise from this change. This fix was thought of to affect as little things as possible.

### CODE REVIEW

**Review by arthurp — CHANGES_REQUESTED** [review]
This change isn't safe in general because `_Value` could be larger than 64-bits. 

What problem are you trying to solve here? The comment you provide really doesn't explain it, because you are just reiterating the fundamentally unsafe nature of unchecked math and those issues apply regardless of the bit width of the types involved.

I get the sense you are trying to avoid a common case of having math with a `DifferenceType` and a `size_t` from elsewhere. However, that's never safe even if the compiler allows it because `DifferenceType` could be any size or type. The user of it doesn't know. So a much better solution is to fix the code to always use `DifferenceType` for the proper ID class whenever it is actually storing differences between those IDs. (That's what `DifferenceType` is for.)

Can you give more details about what the problem you encountered was and the semantics of the types involved (not their representation, their semantics)?

(PS: The format of your comment matches nothing else in the code base. Please use Google C++ Style and doxygen comments. Also I would encourage editing your english prose in the comments a bit. ;-) )

**Comment by GioLomia** [comment]
The issue I am running into has to deal with transition to OpaqueIDLinear. Essentially since we are trying to wrap local and global edge IDs within the OpaqueID class, we explicitly specify that the difference type is _Value, that is than made signed, resulting into a signed 32 bit int. (I might be misunderstanding this part). Which than is in turn compared to an unsigned int in the do_all function. The idea behind this fix was to allow this comparison by increasing the width of the difference type.

**Comment by arthurp** [comment]
Where is do_all doing that comparison? and how are you calling do_all to trigger that behavior? I don't see why do_all would need to do comparisons to incompatible types. If it does, we should look at fixing that instead.

**Comment by GioLomia** [comment]
Within the Executor_DoAll.h on line 131 we compare `m_size <= chunk_size` 
The function:
```
  private:
    bool getWork(Iter& priv_beg, Iter& priv_end, const unsigned chunk_size) {
      bool succ = false;

      work_mutex.lock();
      {
        if (hasWorkWeak()) {
          succ = true;

          Iter nbeg = shared_beg;
          if (m_size <= chunk_size) {
            nbeg = shared_end;
            m_size = 0;

          } else {
            std::advance(nbeg, chunk_size);
            m_size -= chunk_size;
            KATANA_LOG_DEBUG_ASSERT(m_size > 0);
          }

          priv_beg = shared_beg;
          priv_end = nbeg;
          shared_beg = nbeg;
        }
      }
      work_mutex.unlock();

      return succ;
    }
```
I am calling the do_all function from AggregatedGraphView.cpp

```
void
AddMastersToCSR(
    const std::vector<std::unique_ptr<katana::PartitionedGraphView>>&
        component_partitions,
    const std::vector<int64_t>& node_property_offsets,
    std::vector<katana::GlobalNodeID>* local_to_global,
    std::vector<katana::LocalEdgeID>* csr_index,
    katana::NUMAArray<katana::LocalNodeID>* node_property_indices,
    katana::DynamicBitset* deleted_masters) {
  // Create new local ids for all masters we are inheriting. This loop
  // implicitly assumes that the masters of each component are disjoint.
  katana::LocalNodeID new_lid_offset;
  for (size_t i = 0; i < component_partitions.size(); ++i) {
    katana::do_all(
        katana::iterate(
            katana::LocalNodeID{0}, component_partitions[i]->NumMasters()),
        [&](katana::LocalNodeID old_lid) {
          katana::GlobalNodeID gid =
              component_partitions[i]->LocalToGlobalID(old_lid);
          katana::LocalNodeID new_lid(new_lid_offset.value() + old_lid.value());
          local_to_global->at(new_lid.value()) = gid;
          node_property_indices->at(new_lid.value()) =
              component_partitions[i]->GetNodePropertyIndex(old_lid) +
              node_property_offsets[i];
          csr_index->at(new_lid.value()) = katana::LocalEdgeID(std::distance(
              component_partitions[i]->Edges(old_lid).begin(),
              component_partitions[i]->Edges(old_lid).end()));
          if (component_partitions[i]->NodeIsDeleted(old_lid)) {
            deleted_masters->set(new_lid.value());
          }
        },
        katana::steal(), katana::no_stats());

    // no need to check for overflow, we confirmed above that the total number
    // of masters of all the component_partitions is small enough
    new_lid_offset += component_partitions[i]->NumMasters().value();
  }
  ResortVectors(
      katana::LocalNodeID(0), new_lid_offset, local_to_global, csr_index, node_property_indices);
}

```
I did not make any adjustments to the function, besides adapting it for the OpaqueID wraper. I believe the issue comes from the explicit declaration of the difference_type's type (for a lack of a better word). While previously the compiler handled the comparison, it is unable to do so because we explicitly wrap it within signed 32 bit int. (Again, I might be misunderstanding, so please correct me if I am wrong).

**Comment by arthurp** [comment]
Thanks! Now I'm starting to see a more complete picture!

I would argue that this is in fact a bug/oversight in `do_all`. I say this because `chunk_size` is always small (like 64 or maybe 256), and even then clamping it to fit in a smaller type wouldn't hurt anything because it's only a performance tuning parameter and if the number of elements is small then the chunk size doesn't matter. So I think the best fix would be to cast `chunk_size` (with saturation) to `Iter::difference_type` while the loop is being setup and then use that type throughout.

(CC: @gurbinder533 @roshandathathri in case either of you want to chime in about the safety reasonableness of this change.)

**Comment by scober** [comment]
One thing worth mentioning is that I believe it is incorrect for `OpaqueIDLinear::DifferenceType` to be the signed variant of `_Value`. Because it won't be able to represent all possible differences between two `_Value`s.

**Comment by arthurp** [comment]
> One thing worth mentioning is that I believe it is incorrect for `OpaqueIDLinear::DifferenceType` to be the signed variant of `_Value`. Because it won't be able to represent all possible differences between two `_Value`s.

Interesting point. Especially since wrap around semantics in C++ are broken so we cannot rely on them. We could use "one size larger" but that's probably not great because using a 128-bit difference for 64-bit IDs is a bit silly.

Using the unsigned variant of `_Value` doesn't help either because then we cannot represent negative distances which seems important. Probably for this reason, `difference_type` on iterators is signed in STL and I think we should follow that.

So I'm not sure how to deal with this, other than to document that like any unsigned index value, the difference between two indexes is only meaningful if the maximum index used is (true maximum allowed index)/2. This isn't specific to OpaqueID. It also happens with `size_t` vs `ptrdiff_t` for instance.

**Comment by arthurp** [comment]
> It also happens with `size_t` vs `ptrdiff_t` for instance.

This is actually specified: (https://en.cppreference.com/w/cpp/types/ptrdiff_t)

> If an array is so large (greater than PTRDIFF_MAX elements, but less than SIZE_MAX bytes), that the difference between two pointers may not be representable as std::ptrdiff_t, the result of subtracting two such pointers is undefined. 

So I'm not worried about this issue. It's pretty universal in C++ I think.

**Comment by scober** [comment]
I was also looking at that `ptrdiff_t` issue, but I am less willing to accept that limitation for 32 bit integers. Overflowing `ptrdiff_t` requires two pointers that are pretty far apart. But we regularly use more than half of the allowable `LocalNodeID` space. And I would rather not move to 64 bit local node ids.

**Comment by scober** [comment]
Arthur, would you be willing to set a 64 bit floor on the difference type?

**Comment by arthurp** [comment]
> Arthur, would you be willing to set a 64 bit floor on the difference type?

Yes. That's reasonable.

We could even get more tricky and use "one size up" for types smaller than 64-bit and the same same for 64-bit and above. That way we don't use 64-bit differences for 16-bit IDs. But that is probably over engineering it.

**Comment by scober** [comment]
I didn't join Katana to _under_-engineer things!

In all seriousness, I am happy as long as I can take differences in the whole 32 bit range.

File: libsupport/include/katana/OpaqueID.h
Comment (roshandathathri) [inline libsupport/include/katana/OpaqueID.h]: Followed the discussion. Agree with the change. We can stick to 64-bit signed integer as the difference type for any integer type (I'm assuming this is not for float types).

Please fix the comments. Use `///` for doxygen. Drop the excess `!!!..`. Also, add a link to `ptrdiff_t` from the PR in the doxygen code comments.

**Review by roshandathathri — APPROVED** [review]
LGTM but please wait for others' approval before merging.

**Review by arthurp — CHANGES_REQUESTED** [review]
Please implement either the over or under engineered idea from https://github.com/KatanaGraph/katana/pull/384#issuecomment-883752707

I'm not comfortable with using a fixed size like this, but setting a floor or handling smaller and larger types slightly differently is OK. Just make sure to document it on the class since that's where a user would look since they would be passing a template parameter to the class and also on the `DifferenceType` member (using a "see the class" link ideally).

**Comment by GioLomia** [comment]
I like the idea of one size up for types under 64, I just want to clarify one thing. Are we assuming we will not encounter 16 bits because otherwise 16 < 64 and 16++ is 32, which would bring us to the same problem we had before.

**Comment by arthurp** [comment]
> I like the idea of one size up for types under 64, I just want to clarify one thing. Are we assuming we will not encounter 16 bits because otherwise 16 < 64 and 16++ is 32, which would bring us to the same problem we had before.

We are not assuming that. However, I don't think you'll have this issue with a 16-bit ID and a 64-bit difference. Since the issue was related to insufficient range in the difference type not in the ID type. So using a 16-bit ID with a 64-bit difference should be fine and only waste a little space, and since we generally don't have arrays of differences, it shouldn't even matter.

**Comment by scober** [comment]
If we don't modify `do_all` I think we will have the same problem there with the one-size-up solution.

**Comment by arthurp** [comment]
> If we don't modify `do_all` I think we will have the same problem there with the one-size-up solution.

Hmm. Yea. I guess we might. I'm in favor of fixing do_all. Having a fixed type there is clearly wrong in my eyes.

**Comment by ddn0** [comment]
Given the choice of more template metaprogramming or something simple, I would opt for the simple thing, which is to define difference type as `ptrdiff_t` or a typed version thereof for all 64-bit or smaller IDs.

In any event, please also update the documentation where it refers to "signed variant" and consider adding tests to confirm that common instantiations compile / have the correct (lack of) overflow behavior.

**Comment by scober** [comment]
> Given the choice of more template metaprogramming or something simple, I would opt for the simple thing, which is to define difference type as `ptrdiff_t` or a typed version thereof for all 64-bit or smaller IDs.

Does `ptrdiff_t` have advantages over `int64_t`? It seems like if we are going to tie the use of this type to the width of the `_Value` type it would be nice for it to have a clear and known width itself?

**Comment by ddn0** [comment]
> Does ptrdiff_t have advantages over int64_t? It seems like if we are going to tie the use of this type to the width of the _Value type it would be nice for it to have a clear and known width itself?

`ptrdiff_t` is the traditional type for "something that can hold any difference". It also has behavior guaranteed by the standard: https://en.cppreference.com/w/cpp/types/ptrdiff_t . If you want to use `int64_t` please add a static_assert that checks it is at least the size of `ptrdiff_t`.

**Review by roshandathathri — APPROVED** [review]
LGTM

**Review by arthurp — CHANGES_REQUESTED** [review]
@ddn0 said:
> I would opt for the simple thing, which is to define difference type as ptrdiff_t or a typed version thereof for all 64-bit or smaller IDs.

I still want this to be conditional, so as not to limit ourselves to 64-bit IDs.

I'm a bit confused how we ended up back at a fixed type here. We discussed why that's not safe in general.

If people really feel strongly that we don't want to do template metaprogramming here, we could add a static assert that prevents `OpaqueIDLinear` from being used with any type larger than `ptrdiff_t` and use `ptrdiff_t` here and just fix it later if/when we need to. Maybe we will never need IDs larger than 64-bit. We just need to be explicit about that choice and enforce it with static_assert.

**Comment by ddn0** [comment]
> > I would opt for the simple thing, which is to define difference type as ptrdiff_t or a typed version thereof for all 64-bit or smaller IDs.
> 
> I still want this to be conditional, so as not to limit ourselves to 64-bit IDs.
> 
> I'm a bit confused how we ended up back at a fixed type here. We discussed why that's not safe in general.

It is possible to support OpaqueLinearIDs for any ~abstract algebra~ (edit: affine space) but I don't think it is wise to do so. And just because we can use template metaprogramming doesn't mean we should use it. Metaprogramming actually made it easy to make this bug in the first place.

I'd prefer we keep things simple unless there really is a concrete need to do otherwise. If we can fail obviously when things are unexpected, I don't see any harm with picking a specific type.

**Comment by arthurp** [comment]
> . If we can fail obviously when things are unexpected, I don't see any harm with picking a specific type.

Agreed. Sorry I pushed for too much generality. Bad habit of mine.

Let's add a static assert checking the size of `_Value` and move on.

**Comment by GioLomia** [comment]
Just to clarify, which types should cause an exception here?

File: libsupport/include/katana/OpaqueID.h
Comment (arthurp) [inline libsupport/include/katana/OpaqueID.h]: I don't think you need this include. Please remove.

File: libsupport/include/katana/OpaqueID.h
Comment (arthurp) [inline libsupport/include/katana/OpaqueID.h]: Use ptrdiff_t for what? Probably better to say "OpaqueIDLinear only supports values up to the size of ptrdiff_t"

**Review by arthurp — APPROVED** [review]

**Comment by GioLomia** [comment]
Just squashed the commits together to make the history cleaner.

**Pull request merged**
